File: vscode-extension/src/service/CompletionService.ts
/**
 * 代码补全服务
 */

import * as vscode from 'vscode';
import { ConfigService } from '../config/index.js';
import { ContextService } from './ContextService.js';

/**
 * 补全项
 */
interface CompletionItem {
    text: string;
    startLine: number;
    confidence: number;
}

/**
 * 代码补全提供者
 */
export class CompletionProvider implements vscode.InlineCompletionItemProvider {
    
    private config: ConfigService;
    private contextService: ContextService;
    
    constructor(config: ConfigService) {
        this.config = config;
        this.contextService = new ContextService();
    }
    
    /**
     * 提供行内补全
     */
    async provideInlineCompletionItems(
        document: vscode.TextDocument,
        position: vscode.Position,
        context: vscode.InlineCompletionContext,
        token: vscode.CancellationToken
    ): Promise<vscode.InlineCompletionList | vscode.InlineCompletionItem[]> {
        if (!this.config.isConfigured()) {
            return [];
        }
        
        try {
            // 获取当前行之前的上下文
            const startLine = Math.max(0, position.line - 20);
            const range = new vscode.Range(startLine, 0, position.line, position.character);
            const contextText = document.getText(range);
            
            // 调用 LLM 获取补全
            const completions = await this.getCompletions(contextText, document.languageId);
            
            return completions.map(c => new vscode.InlineCompletionItem(
                c.text,
                new vscode.Range(position, position)
            ));
        } catch {
            return [];
        }
    }
    
    /**
     * 获取补全
     */
    private async getCompletions(context: string, language: string): Promise<CompletionItem[]> {
        const prompt = `请根据上下文预测最可能的代码补全。

语言: ${language}
上下文:
${context}

请只返回补全的代码，不需要解释。`;
        
        const response = await fetch(`${this.config.getApiEndpoint()}/v1/chat/completions`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${this.config.getApiKey()}`,
            },
            body: JSON.stringify({
                model: this.config.getModel(),
                messages: [
                    { role: 'system', content: '你是一个代码补全助手，只返回补全代码，不做解释。' },
                    { role: 'user', content: prompt },
                ],
                max_tokens: 256,
                temperature: 0.3,
            }),
        });
        
        if (!response.ok) {
            return [];
        }
        
        const data = await response.json();
        const content = data.choices?.[0]?.message?.content;
        
        if (!content) {
            return [];
        }
        
        // 清理代码（移除 markdown 代码块标记）
        const cleanedCode = content
            .replace(/^```[\w]*\n?/, '')
            .replace(/\n?```$/, '')
            .trim();
        
        if (!cleanedCode) {
            return [];
        }
        
        return [{
            text: cleanedCode,
            startLine: 0,
            confidence: 0.85,
        }];
    }
}

File: idea-plugin/bin/main/com/vex/coding/service/CompletionService.kt
package com.vex.coding.service

import com.vex.coding.config.ConfigService
import kotlinx.coroutines.*
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * 代码补全项
 */
data class CompletionItem(
    val text: String,
    val startLine: Int,
    val confidence: Double
)

/**
 * 代码补全服务
 */
class CompletionService {
    
    private val config = ConfigService.getInstance()
    private val contextService: ContextService
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(5))
        .build()
    
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    constructor(project: com.intellij.openapi.project.Project) {
        this.contextService = ContextService(project)
    }
    
    /**
     * 获取补全建议
     */
    suspend fun getCompletions(
        file: com.intellij.openapi.vfs.VirtualFile,
        position: com.intellij.openapi.editor.LogicalPosition
    ): List<CompletionItem> = withContext(Dispatchers.IO) {
        if (!config.isConfigured()) {
            return@withContext emptyList()
        }
        
        try {
            val context = contextService.buildContext()
            val source = context.files.firstOrNull()?.content ?: return@withContext emptyList()
            
            // 提取光标周围的代码
            val lines = source.split("\n")
            val aroundCursor = lines.drop(
                maxOf(0, position.line - 10)
            ).take(20).joinToString("\n")
            
            val request = buildCompletionRequest(aroundCursor, file.extension ?: "txt")
            
            val httpRequest = HttpRequest.newBuilder()
                .uri(URI.create("${config.getApiEndpoint()}/v1/chat/completions"))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer ${config.getApiKey()}")
                .POST(HttpRequest.BodyPublishers.ofString(request))
                .timeout(Duration.ofSeconds(10))
                .build()
            
            val response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString())
            
            if (!response.statusCode().toString().startsWith("2")) {
                return@withContext emptyList()
            }
            
            parseCompletionResponse(response.body())
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * 构建补全请求
     */
    private fun buildCompletionRequest(context: String, language: String): String {
        val prompt = """请根据上下文预测最可能的代码补全。

语言: $language
上下文:
```
$context
```

请只返回补全的代码，不需要解释。"""
        
        return buildString {
            append("{")
            append("\"model\": \"${config.getModel()}\",")
            append("\"messages\": [")
            append("{\"role\": \"system\", \"content\": \"你是一个代码补全助手，只返回补全代码，不做解释。\"},")
            append("{\"role\": \"user\", \"content\": ${prompt.jsonEscape()}}")
            append("],")
            append("\"max_tokens\": 256,")
            append("\"temperature\": 0.3")
            append("}")
        }
    }
    
    /**
     * 解析补全响应
     */
    private fun parseCompletionResponse(body: String): List<CompletionItem> {
        val contentStart = body.indexOf("\"content\":\"") + 11
        val contentEnd = body.indexOf("\"", contentStart)
        
        if (contentStart > 10 && contentEnd > contentStart) {
            val content = body.substring(contentStart, contentEnd)
                .replace("\\n", "\n")
                .replace("\\\"", "\"")
                .replace("```", "")
                .trim()
            
            if (content.isNotBlank()) {
                return listOf(
                    CompletionItem(
                        text = content,
                        startLine = 0,
                        confidence = 0.85
                    )
                )
            }
        }
        
        return emptyList()
    }
    
    private fun String.jsonEscape(): String {
        return this.replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
    }
}